### ISSUE TITLE
new base network trials

### ISSUE THREAD

**Issue by youngwanLEE** [weiliu89/caffe#34]
Thanks in advance for your wonderful code :)

Recently, I tried to change the VGG-16 base network to ResNet-50L & ResNet-101L by using pretrained model from K. He Github repo. and reffering your code 'ResBody' and 'Res152Body' in model_libs.py.

In ResNet-50L version, I used res3d_branch2c and res4f_branch2c for mbox_source_layers.
In ResNet-101L version, I used res3b3_branch2c and res4b22_branch2c for mbox_source_layers.

In addition, I also freezed relatively previous layers.

However, the results on ResNet50 & ResNet101 always were not better than VGG-base.

I read in previous Issues that you already successfully tested Inception-v3 and ResNet101L. I wonder your results.

And Could you uploaded your pretrained model on Inception-v3 and ResNet101L ?

Thank you.

**Comment by weiliu89** [comment]
There were some issues for the first few layers of ResNet152Body, I have fixed that in b7b3929. Sorry for not fixing it earlier.

I get similar (a little worse actually) results using ResNet101 as of VGG16 on VOC0712. But if you check [Table 10](https://arxiv.org/pdf/1512.03385v1.pdf), the pure improvement of using ResNet101 is not very high. There are some details how to train RPN with ROI layer in page 10.

In the [new paper](https://arxiv.org/pdf/1605.06409v2.pdf), on page 5, the result is very similar to the number of VGG16, if not using the atrous trick.

For Inception-v3, I converted the TF model to caffe format, but I couldn't get similar classification number as reported. In specific, my converted Inception v3 only has 65% top5 accuracy, which is low. Probably there are some problems of the converted model, and I don't have time to examine the problem. The detection result is very close (~69.8) to VGG and ResNet, but the advantage is that it is faster to train and use less memory compared to ResNet101.

I am curious what numbers do you get?

**Comment by youngwanLEE** [comment]
Actually, I finetuned using KITTI dataset for road object detection.

The test results were shown as below : 
val_acc : VGG - 0.39 | Res101 - 0.34 | Res50 - 0.33

I have another issue : processing time

I experimented processing times of various version using KITTI test dataset : 2494 images

I tried to use two metrics : 1) python script reffered to ipynb  2)./build/tools/caffe time  

-------------------------------------------------------------------------------------------------------------------
1) python script

I reffered to your ssd_detect.ipynb and time check like this : 

for s in image_list :
...
...
    # Forward pass.
    start = time.time()
    detections = net.forward()['detection_out']
    end = time.time() - start
    sumPTime += end
    ....
    ...
avgPTime = sumPTime/cntFrame

With this metric, The results are :

On 980 Ti,        VGG - 22ms | Res50 - 32ms
On TX1 board, VGG-  1.01s | Res50 - 336ms


-------------------------------------------------------------------------------------------------------------------

2) caffe time --model = deploy.prototxt

On 980 Ti,        VGG - 22.75ms | Res50 - 24.72ms
On TX1 board, VGG-  262.931ms | Res50 - 229.372ms


-------------------------------------------------------------------------------------------------------------------

As you can see, on TX1 board experimental results are strange. 

VGG result using python script on TX1 is very slow whereas on 980Ti result  is normal. 

Q1. 
I wonder why VGG version on TX1 board is slower than Res50L one.

Q2.
In addition, without test images 'caffe time command' measured the whole network before detection_output layer which computed NMS.

for the exact time-check, when using 'caffe time command',  should we have included detection_output layers ? 

Actually, I tried to caffe time command with --model=test.prototxt, But,

Unknown bottom blob 'data' (layer 'conv1', bottom index 0)
*** Check failure stack trace: ***

this error message occurred because of data layer.

In conclusion,  I want to know exactly evaluate FPS(or processing time ) on the SSD framework. 

Thank you very much

**Comment by youngwanLEE** [comment]
Thanks for your reply. 

I commented out the line and executed score_ssd_pascal.py, however, still only loss and detection_eval appeared.

Which one I mistake?

**Comment by weiliu89** [comment]
You can get the time interval and estimate the FPS. For example, I got the following lines after I run `examples/ssd/score_ssd_pascal.py`:

I0713 **20:50:25.756376** 117599 net.cpp:684] Ignoring source layer mbox_loss
I0713 **20:51:32.472930** 117599 solver.cpp:531]     Test net output #0: detection_eval = 0.723217
I0713 20:51:32.473057 117599 solver.cpp:325] Optimization Done.

The total time spent on evaluating 4952 test images is about 67 seconds, and thus FPS = 4952 / 67 = 74. Actually the time also includes evaluating the detection results, but it can be ignored since it is fast.

**Comment by youngwanLEE** [comment]
Thank you very much! :)

**Comment by ck196** [comment]
@weiliu89 can you share scripts to train inception-v3 like ssd_pascal.py script?

**Comment by weiliu89** [comment]
@ck196 You could refer to the ResNet script.
